src/tools/tools/process.fs

```diff
@@ -1171,10 +1171,97 @@ type internal ThreadPoolWait internal (id:int) as this =
         member x.Dispose() = 
             CleanUp.Current.CleanUpAll()
             GC.SuppressFinalize(x)
-            
-
-            
-
+          
+/// ThreadPoolWait.RegisterWaitForSingleObject, with capability:
+/// 1. To remove a particular wait handle. 
+/// 2. To show all wait handles queued in the system, 
+/// and 3. To show long continuation (not normal). 
+type internal ThreadPoolWait() = 
+    static let rwhToRemove = ConcurrentQueue<_>()
+    static let rwhLookup = ConcurrentDictionary<_,_>()
+    static let waitHandleCollection = ConcurrentDictionary<_,_>() 
+    static let unregisterTimer = new System.Threading.Timer( TimerCallback( ThreadPoolWait.ToUnregister ), null, 10, 10 )
+    static member WaitForHandleWithCancellation (infoFunc: unit-> string) (handle:WaitHandle) (continuation:unit->unit) (unblockHandle:ManualResetEvent) (token:CancellationToken) =
+        token.Register( fun _ -> ThreadPoolWait.TryWakeup handle |> ignore ) |> ignore 
```

Comment:
(Even with the check I mentioned above) it is possible that we register the "TryWakeUp" with the token,  but a cancellation request happens before we move to the WaitForHandle (or in mid of it), the TryWakeUp is called and did nothing, and we go ahead with WaitForHandle.   The point is:  register TryWakeUp with token does not ensure it will indeed do the wakeup.  So we cannot completely rely on it.  Do we have any other measures? If we do, is this  "register" still useful?